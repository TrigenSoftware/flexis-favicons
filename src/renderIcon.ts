import {
	promisify
} from 'util';
import Vinyl from 'vinyl';
import Sharp from 'sharp';
import svg2imgCb from 'svg2img';
import {
	ISize
} from './icons';
import {
	isSvg
} from './extensions';
import {
	attachMetadata
} from './helpers';

interface IRenderConfig {
	rotate: boolean;
	width: number;
	height: number;
	background: string;
	offset: number;
}

const PERCENTS_100 = 100;
const TWICE = 2;
const ROTATE_DEGREES = 90;

const svg2img = promisify(svg2imgCb);

/**
 * Render icon.
 * @param  sources - Array of sources.
 * @param  renderConfig - Render config.
 * @return Rendered icon.
 */
export default async function renderIcon(sources: Vinyl[], {
	rotate,
	width,
	height,
	background,
	offset
}: IRenderConfig) {

	const maximumSide = Math.max(width, height);
	const offsetPx = Math.round(maximumSide / PERCENTS_100 * offset) || 0;
	const canvas = await createCanvas(width, height, background);
	const sprite = await createSprite(sources, rotate, width, height, offsetPx);
	const renderedSprite = await sprite.png().toBuffer();

	canvas.composite([{
		input: renderedSprite
	}]);

	return canvas.png().toBuffer();
}

/**
 * Create canvas for rendering.
 * @param  width - Width of canvas.
 * @param  height - Height of canvas.
 * @param  background - Background color of canvas.
 * @return Canvas.
 */
function createCanvas(
	width: number,
	height: number,
	background: string
) {
	return Sharp(undefined, {
		create: {
			width,
			height,
			background,
			channels: 4
		}
	});
}

/**
 * Create sprite.
 * @param  sources - Sprite sources.
 * @param  rotate - Rotate 90deg or not.
 * @param  width - Width of sprite.
 * @param  height - Height of sprite.
 * @param  offset - Offset from canvas edges.
 * @return Sprite.
 */
async function createSprite(
	sources: Vinyl[],
	rotate: boolean,
	width: number,
	height: number,
	offset: number
) {

	const spriteWidth = width - offset * TWICE;
	const spriteHeight = height - offset * TWICE;
	const source = await getSuitableSourceBuffer(sources, spriteWidth, spriteHeight);
	const sprite = Sharp(source);

	if (rotate) {
		sprite.rotate(ROTATE_DEGREES);
	}

	sprite.resize(spriteWidth, spriteHeight, {
		fit: 'inside'
	});

	return sprite;
}

/**
 * Get suitable source by size.
 * @param  sources - Array of sources.
 * @param  width - Needed width.
 * @param  height - Needed height.
 * @return Siutable source.
 */
async function getSuitableSourceBuffer(
	sources: Vinyl[],
	width: number,
	height: number
): Promise<Buffer> {

	const svgSource = sources.find(({ basename }) => isSvg(basename));

	if (svgSource) {

		await attachMetadata(svgSource);

		const icon: Buffer = await svg2img((svgSource.contents as Buffer).toString('utf8'), {
			preserveAspectRatio: 'xMidYMid meet',
			...getContainSize({
				width,
				height
			}, svgSource.metadata)
		});

		return icon;
	}

	for (const source of sources) {
		await attachMetadata(source);
	}

	const maximumSide = Math.max(width, height);
	const nearestIcon = sources.reduce((nearestIcon, source) => {

		const {
			width: nearestIconWidth,
			height: nearestIconHeight
		} = nearestIcon.metadata;
		const {
			width: sourceWidth,
			height: sourceHeight
		} = source.metadata;
		const nearestIconMaximumSide = Math.max(nearestIconWidth, nearestIconHeight);
		const sourceMaximumSide = Math.max(sourceWidth, sourceHeight);

		if (
			(nearestIconMaximumSide > sourceMaximumSide || nearestIconMaximumSide < maximumSide)
			&& sourceMaximumSide >= maximumSide
		) {
			return source;
		}

		return nearestIcon;
	}, sources[0]);

	return nearestIcon.contents as Buffer;
}

/**
 * Get size of icon to contain in box.
 * @param boxSize - Size of box.
 * @param iconsSize - Size of icon.
 */
function getContainSize(
	{
		width: boxWidth,
		height: boxHeight
	}: ISize,
	{
		width: iconWidth,
		height: iconHeight
	}: ISize
): ISize {

	const minBoxSide = Math.min(boxWidth, boxHeight);
	const isWidthMin = minBoxSide === boxWidth;

	return isWidthMin ? {
		width:  boxWidth,
		height: Math.round(iconHeight / iconWidth * boxWidth)
	} : {
		height: boxHeight,
		width:  Math.round(iconWidth / iconHeight * boxHeight)
	};
}
