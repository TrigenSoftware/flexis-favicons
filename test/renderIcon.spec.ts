import fs from 'fs';
import path from 'path';
import renderIcon from '../src/renderIcon';
import {
	svg,
	png
} from './favicon';

describe('renderIcon', () => {

	it('should render correct icon from svg', async () => {

		const icon = await renderIcon([svg], {
			rotate:     false,
			width:      80,
			height:     80,
			background: 'red',
			offset:     20
		});

		fs.writeFileSync(path.join(__dirname, 'renderIcon_svg.png'), icon);
		expect(JSON.stringify(icon)).toMatchSnapshot();
	});

	it('should render correct icon from png', async () => {

		const icon = await renderIcon([png], {
			rotate:     false,
			width:      80,
			height:     80,
			background: 'red',
			offset:     20
		});

		fs.writeFileSync(path.join(__dirname, 'renderIcon_png.png'), icon);
		expect(JSON.stringify(icon)).toMatchSnapshot();
	});

	it('should render correct rotated icon from svg', async () => {

		const icon = await renderIcon([svg], {
			rotate:     true,
			width:      128,
			height:     128,
			background: 'blue',
			offset:     25
		});

		fs.writeFileSync(path.join(__dirname, 'renderIcon_svg_rotated.png'), icon);
		expect(JSON.stringify(icon)).toMatchSnapshot();
	});

	it('should render correct rotated icon from png', async () => {

		const icon = await renderIcon([png], {
			rotate:     true,
			width:      128,
			height:     128,
			background: 'blue',
			offset:     25
		});

		fs.writeFileSync(path.join(__dirname, 'renderIcon_png_rotated.png'), icon);
		expect(JSON.stringify(icon)).toMatchSnapshot();
	});
});
