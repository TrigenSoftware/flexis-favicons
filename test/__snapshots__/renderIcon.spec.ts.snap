// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renderIcon should render better quality from svg source 1`] = `"576b4253aacd3eea97873366db538504348c306e"`;

exports[`renderIcon should render better quality from svg source 2`] = `"c2826288f9c81d0a83c51ce4a9b868d3428d5a0a"`;

exports[`renderIcon should render correct icon from png 1`] = `"74cab8efa35abcb9b6789cbf395f9ca2bc71012a"`;

exports[`renderIcon should render correct icon from svg 1`] = `"76026a4d85751a41a803bc988c188a315b063fe2"`;

exports[`renderIcon should render correct rotated icon from png 1`] = `"5969a63e0deee15253582c4b9b90778f0f667d8e"`;

exports[`renderIcon should render correct rotated icon from svg 1`] = `"8ba13a98231c7d00e969682383351b826a69dc8b"`;

exports[`renderIcon should render correct rotated non-square icon from png 1`] = `"b923c1bc735ed26b20b8b30382bfa58479da22d7"`;

exports[`renderIcon should render correct rotated non-square icon from svg 1`] = `"7f6b692d6228dd85e8a63bb3ec3e762820477fe8"`;
